src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -396,6 +397,50 @@ private void VerifyIsAlive()
                 throw new InvalidOperationException(SR.AssemblyLoadContext_Verify_NotUnloading);
             }
         }
+
+        private static readonly AsyncLocal<AssemblyLoadContext> _asyncLocalCurrentContextualReflectionContext = new AsyncLocal<AssemblyLoadContext>();
+
+        public static AssemblyLoadContext CurrentContextualReflectionContext
+        {
+            get { return _asyncLocalCurrentContextualReflectionContext.Value; }
+        }
+
+        public ContextualReflectionScope EnterContextualReflection()
+        {
+            return new ContextualReflectionScope(this);
+        }
+
+        static public ContextualReflectionScope EnterContextualReflection(Assembly activating)
+        {
+            return activating != null ?
+                GetLoadContext(activating).EnterContextualReflection() :
+                new ContextualReflectionScope(null);
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public struct ContextualReflectionScope : IDisposable
+        {
+            private bool disposed;
+            private AssemblyLoadContext predecessor;
+
+            public ContextualReflectionScope(AssemblyLoadContext activating)
+            {
+                predecessor = AssemblyLoadContext.CurrentContextualReflectionContext;
+                disposed = false;
+                AssemblyLoadContext._asyncLocalCurrentContextualReflectionContext.Value = activating;
+            }
+
+            public void Dispose()
+            {
+                if(!this.disposed)
+                {
+                    AssemblyLoadContext._asyncLocalCurrentContextualReflectionContext.Value = predecessor;
+                    disposed = true;
+                }
+            }
+        }
+
+        internal IntPtr GetNativeAssemblyLoadContext() => _nativeAssemblyLoadContext;
```

Comment:
Does this need to be managed reference to make sure that it is not unloaded while in use?